sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs

```diff
@@ -0,0 +1,188 @@
+            {
+                RunningTaskTokenSource = new CancellationTokenSource();
+
+                InnerConsumer = InnerClient.CreateConsumer(ConsumerGroup, PartitionId, Options.InitialEventPosition);
+
+                await PartitionProcessor.Initialize().ConfigureAwait(false);
+
+                RunningTask = Run(RunningTaskTokenSource.Token);
+            }
+        }
+
+        /// <summary>
+        ///   Stops the partition pump.  In case it hasn't been started, nothing happens.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public async Task Stop() => await Stop(PartitionProcessorCloseReason.Shutdown);
+
+        /// <summary>
+        ///   Stops the partition pump.  In case it hasn't been started, nothing happens.
+        /// </summary>
+        ///
+        /// <param name="reason">The reason why the partition pump is being closed.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        private async Task Stop(PartitionProcessorCloseReason reason)
+        {
+            if (RunningTask != null)
+            {
+                RunningTaskTokenSource.Cancel();
+                RunningTaskTokenSource = null;
+
+                await RunningTask;
+                RunningTask = null;
+
+                await InnerConsumer.CloseAsync();
+                InnerConsumer = null;
+
+                await PartitionProcessor.Close(reason).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        ///   The main loop of a partition pump.  It receives events from the Azure Event Hubs service
+        ///   and delegates their processing to the inner partition processor.
+        /// </summary>
+        ///
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        private async Task Run(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                IEnumerable<EventData> receivedEvents = null;
+
+                try
+                {
+                    receivedEvents = await InnerConsumer.ReceiveAsync(Options.MaximumMessageCount, Options.MaximumReceiveWaitTime);
+                }
+                catch (Exception exception)
+                {
+                    await PartitionProcessor.ProcessError(exception).ConfigureAwait(false);
+
+                    // Stop the pump if it's not a known retryable exception.
+
+                    if (!(exception is EventHubsException) || !(exception as EventHubsException).IsTransient)
+                    {
+                        _ = Stop(PartitionProcessorCloseReason.EventHubException);
```

Comment:
Why are you not awaiting or returning this task?    You've introduced some non-determinism and potential for exceptions to go unobserved.     You'll also want `ConfigureAwait(false)` here, if you're awaiting.

Reply:
We can't await it because `Stop` awaits this method, so we would have a deadlock. I agree that returning it is a better idea, so that's what I'll do.

Reply:
Just added a comment explaining the situation, but the original behavior was kept.